FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1148 @@
+
+        Ref CmpNode {};
+        if (ReducedPrecisionMode) {
+          CmpNode = IREmit->_FCmp(8, StackValue1, StackValue2);
+        } else {
+          CmpNode = IREmit->_F80Cmp(StackValue1, StackValue2);
+        }
+
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+      case OP_F80STACKTEST: {
+        const auto* Op = IROp->C<IROp_F80StackTest>();
+        auto Offset = Op->SrcStack;
+        auto Value = MigrateToSlowPath_IfInvalid(Offset);
+        auto StackNode = LoadStackValue(Value, Offset);
+        Ref ZeroConst = IREmit->_VCastFromGPR(ReducedPrecisionMode ? 8 : 16, 8, GetConstant(0));
+
+        Ref CmpNode {};
+        if (ReducedPrecisionMode) {
+          CmpNode = IREmit->_FCmp(8, StackNode, ZeroConst);
+        } else {
+          CmpNode = IREmit->_F80Cmp(StackNode, ZeroConst);
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPVALUE: {
+        const auto* Op = IROp->C<IROp_F80CmpValue>();
+        const auto& Value = CurrentIR.GetNode(Op->X80Src);
+        auto StackTop = MigrateToSlowPath_IfInvalid();
+        auto StackNode = LoadStackValue(StackTop);
+
+        Ref CmpNode {};
+        if (ReducedPrecisionMode) {
+          CmpNode = IREmit->_FCmp(8, StackNode, Value);
+        } else {
+          CmpNode = IREmit->_F80Cmp(StackNode, Value);
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80XTRACTSTACK: {
+        auto Value = MigrateToSlowPath_IfInvalid();
+        Ref St0 = LoadStackValue(Value);
+
+        Ref Exp {};
+        Ref Sig {};
+        if (ReducedPrecisionMode) {
+          std::tie(Exp, Sig) = SplitF64SigExp(St0);
+        } else {
+          Exp = IREmit->_F80XTRACT_EXP(St0);
+          Sig = IREmit->_F80XTRACT_SIG(St0);
+        }
+
+        if (SlowPath) {
+          // Write exp to top, update top for a push and set sig at new top.
+          StoreStackValueAtTop_Slow(Exp, false);
+          UpdateTopForPush_Slow();
+          StoreStackValueAtTop_Slow(Sig);
+        } else {
+          StackData.setTop(StackMemberInfo {Exp});
+          StackData.push(StackMemberInfo {Sig});
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_SYNCSTACKTOSLOW: {
+        // This synchronizes stack values but doesn't necessarily moves us off the FastPath!
+        Ref NewTop = SynchronizeStackValues();
+        IREmit->ReplaceUsesWithAfter(CodeNode, NewTop, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_STACKFORCESLOW: {
+        MigrateToSlowPathIf(true);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_INCSTACKTOP:
+      case OP_DECSTACKTOP: {
+        if (SlowPath) {
+          if (IROp->Op == OP_DECSTACKTOP) {
```

Comment:
Oops, I didn't see that there was a push vs pop distinction. I guess I shouldn't've said to combine these, my mistake! 😿